### ISSUE TITLE
Use semantic model

### ISSUE THREAD

**Issue by andry-tino** [andry-tino/Rosetta#19]
Enable the usage of semantic model when translating.

**Rationale** Type references are a problem, the single name is not enough, the full name should be specified.

## Must
Scoped for ScriptSharp definition only.

- Have full names in types for classes and interfaces in base lists
- Have full names in methods, getters, setters, properties for parameters and returned values

## Implementation details
Allow the possibility to specify an assembly as input when processing input files, this would enable the loading of the semantic model.

**Comment by andry-tino** [comment]
## Effort
In order to have this, it is necessary to have `TranslationUnitFactory`s accept `SemanticModel` so they can pass it to `Helper`s.

```
     +------------+
     | AST Walker |
     +------+-----+
            |
            |
+-----------v------------+
| TranslationUnitFactory |
+-----------+------------+
            |
            |
       +----v----+
       | Helper  |
       +---------+
```

**Comment by andry-tino** [comment]
Change 6735f72 has enabled semantic model to be used in the code, however, some parts of the code need to use it as the logic for relying on that was commented out long ago.

**Comment by andry-tino** [comment]
## Problem with transformation
A problem arises concerning syntax transformation.

**What** Issue is that after transforming, the syntax tree is different and does not correspond anymore to the semantic model.

## Solutions
Possible approaches might consider:

- Getting rid of transformation and operate `ScriptNamespace` handling at AST level.
- Keeping the original non transformed tree in memory and having a mapping between nodes from transformed one and original. When asking for symbols in the tree, the mapping is triggered and the node in the old tree is used.

### Mappings
In the mappings case, favorable, the approach should be:

1. Try finding the symbol in the semantic model
2. If not found, look in mappings

**Problem** Mappings are not always good, we might parse single files separately with a common semantic model always loaded from scratch.

### Semantic model update
Semantic model can be updated. See: [Modify syntax tree then get updated semantic model](http://stackoverflow.com/questions/12164769/modify-syntax-tree-then-get-updated-semantic-model).

**Problem** Mappings are not always good, we might parse single files separately with a common semantic model always loaded from scratch.

**Possible solution** However Roslyn provides a method: `GetAttributes` on `Symbol` which returns the attributes. So when doing the transformation, we check first wheter all referenced types in the tree have the `ScriptNamespace` applied, in case, we store a mapping or we edit the semantic model (unfeasible as it must happen through edit of the syntax tree).

## Favorable solution
A mixture of the mapping approach and the semantic model update.

**Available types in AST** At transformation time, we transform the AST and then generate an updated model. This will deal with the types available in the local AST (a small part represented in the semantic model).

**Referenced unavailable types** We still have a problem concerning types referenced by the local AST but not available. The semantic model will refer to them using the not overriden namespace, while we should actually use the overridden one.

To accommodate this, helpers (when provided with semantic model), when providing the full name or every namespace-related quantity, will check the presence of the `ScriptNamespace` attribute and override the name accordingly. This will impact ONLY helpers in the ScriptSharp Definition area.

**Comment by andry-tino** [comment]
The issue has been solved by 	86060aa, which also generated issue #22.
